asset-services/asset-services-celery/src/celery_box.rs

```diff
@@ -0,0 +1,106 @@
+//! Broker-generic interface to [`Celery`].
+//!
+//! This module allows code to refer to a [`Celery`] instance without hard-coding the choice of broker,
+//! using the [`CeleryBox`] type.
+
+use std::io;
+use std::sync::Arc;
+
+use celery::broker::{AMQPBroker, RedisBroker};
+use celery::error::CeleryError;
+use celery::task::{AsyncResult, Signature, Task};
+use celery::Celery;
+use url::Url;
+
+use self::CeleryBox::{Amqp, Redis};
+
+/// Wrap [`Celery`] struct.
+pub enum CeleryBox {
+    Amqp(Arc<Celery<AMQPBroker>>),
+    Redis(Arc<Celery<RedisBroker>>),
+}
+
+impl CeleryBox {
+    /// Build an instance using [`Celery::builder`], with the broker determined by the scheme of `broker_url`.
+    pub async fn build(name: &str, broker_url: &str) -> Result<Self, CeleryError> {
```

Comment:
@PiDelport I see here the broker instance can either be AMQP or Redis. How does it determine which broker instance to initialize? Is this set in the environment variable in docker compose 'BROKER_URL'?

Reply:
Yup, this helper determines it at runtime by looking at the URL scheme of the `broker_url` passed in by `init_celery` / `init_celery_from_env`.

It's not ideal that we need this wrapper, but it makes it possible for us to switch between RabbitMQ and Redis without editing and recompiling all the downstream references every time, which would make testing broker-related things very painful.

This is something that rusty-celery should support better: I plan to file an issue about it.